python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py

```diff
@@ -84,13 +118,169 @@ def build_time_partition_freshness_checks(
         deadline_cron=deadline_cron,
         timezone=timezone,
         severity=severity,
-        asset_property_enforcement_lambda=lambda assets_def: check.invariant(
-            isinstance(assets_def.partitions_def, TimeWindowPartitionsDefinition),
-            "Asset is not time-window partitioned.",
-        ),
     )
+
+
+def _build_freshness_multi_check(
+    asset_keys: Sequence[AssetKey],
+    deadline_cron: str,
+    timezone: str,
+    severity: AssetCheckSeverity,
+) -> AssetChecksDefinition:
+    params_metadata: dict[str, Any] = {
+        TIMEZONE_PARAM_KEY: timezone,
+        DEADLINE_CRON_PARAM_KEY: deadline_cron,
+    }
+
+    @multi_asset_check(
+        specs=[
+            AssetCheckSpec(
+                "freshness_check",
+                asset=asset_key,
+                metadata={FRESHNESS_PARAMS_METADATA_KEY: params_metadata},
+                description="Evaluates freshness for targeted asset.",
+            )
+            for asset_key in asset_keys
+        ],
+        can_subset=True,
+        name=f"freshness_check_{unique_id_from_asset_keys(asset_keys)}",
+    )
+    def the_check(context: AssetCheckExecutionContext) -> Iterable[AssetCheckResult]:
+        for check_key in context.selected_asset_check_keys:
+            asset_key = check_key.asset_key
+            current_timestamp = pendulum.now("UTC").timestamp()
+
+            partitions_def = cast(
+                TimeWindowPartitionsDefinition,
+                context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
+            )
+            check.invariant(
+                isinstance(partitions_def, TimeWindowPartitionsDefinition),
+                "Expected asset to be time-window partitioned",
+            )
+            current_time_in_freshness_tz = pendulum.from_timestamp(current_timestamp, tz=timezone)
+            deadline = get_latest_completed_cron_tick(
+                deadline_cron, current_time_in_freshness_tz, timezone
+            )
+            deadline_in_partitions_def_tz = pendulum.from_timestamp(
+                deadline.timestamp(), tz=partitions_def.timezone
+            )
+            last_completed_time_window = check.not_none(
+                partitions_def.get_prev_partition_window(deadline_in_partitions_def_tz)
+            )
+            expected_partition_key = partitions_def.get_partition_key_range_for_time_window(
+                last_completed_time_window
+            ).start
+
+            last_update_time_lower_bound = check.not_none(
+                last_completed_time_window,
+                "Expected a valid partitioned asset with a completed time window in order to determine valid freshness window.",
+            ).end
+
+            latest_record = retrieve_latest_record(
+                instance=context.instance, asset_key=asset_key, partition_key=expected_partition_key
+            )
+            update_timestamp = get_last_updated_timestamp(latest_record, context)
```

Comment:
I'm confused about why `update_timestamp` is figuring into the partition-based freshness calculation?